### DISCUSSION TITLE
How to serve an app's static files.

### DISCUSSION THREAD

**Discussion by cstrutton** [nginx-proxy/nginx-proxy#1962] [Q&A]
Thanks for all the work you have put into this great project.  I must be missing something, but how do I serve an app's static files out of the same container.  Do I need to create per VHOST config pointing to a shared volume?  Is there an easier way that I am not seeing?

**Comment by jwilder** [comment]
`nginx-proxy` just proxies connections to your containers.  It can't reach into your container to serve up files from the file system automatically.  You would need to have a handler serving up your static assets when `nginx-proxy` forwards those request back to the container.

You could also use a custom template or vhost config pointing to a shared volume as you said so that `nginx-proxy` does have access to the files and can serve them directly.

**Comment by janekolszak** [comment]
#### If you can use _separate VIRTUAl_HOST_/domain for static content (say static.example.com):
- Create _static.example.com.conf_ file with something like:

```
server {
    server_name static.example.com;
    listen 80;
    access_log /var/log/nginx/access.log;
    location / {
        root /www/data;
    }
}
```
- In dockerfile: COPY static.example.com.conf /etc/nginx/conf.d
- Mount/Copy the data to the nginx container

#### If you can't use a separate domain
- Create _example.com_ file:

```
location /static {
    root /www/data;
}
```
- In dockerfile: COPY example.com /etc/nginx/vhost.d/
- Mount/Copy the data to the nginx container

### You may consider a regex for a clever location specification.

**Comment by uLan08** [comment]
Since this issue was never closed....

I was able to serve static files this way
```
server {
    server_name static.example.com;
    listen 80;
    access_log /var/log/nginx/access.log;
    location / {
        root /www/data;
    }
}
```
Will the proxy configurations from `/etc/nginx/conf.d/default.conf` apply to the one I created? If not, are there any repercussions to this?

**Comment by aliallaoui** [comment]
@janekolszak solution is great but could it be done in a more general manner by editing nginx.tmpl and using environment variables ( like STATIC_DOMAIN and STATIC_LOCATION) ? I am not used to go templating language so any help is welcome ;)

**Comment by uLan08** [comment]
For me, I went on a different route than what I originally commented here. I set up a vanilla nginx container that does the serving then I just set it with the usual env vars, `VIRTUAL_HOST` and `VIRTUAL_PORT` and `nginx-proxy` just proxies to that container. This approach adheres to the docker way of containers having only one responsibility and also I think is what jwilder originally suggested.

**Comment by 1fexd** [comment]
> For me, I went on a different route than what I originally commented here. I set up a vanilla nginx container that does the serving then I just set it with the usual env vars, `VIRTUAL_HOST` and `VIRTUAL_PORT` and `nginx-proxy` just proxies to that container. This approach adheres to the docker way of containers having only one responsibility and also I think is what jwilder originally suggested.

I get that this is the "cleanest" way, but isn't a setup like this quite resource-expensive, since every request for a static file has to pass through two nginx instances? I also found [this](https://stackoverflow.com/questions/54571819/serving-static-files-from-jwilder-nginx-proxy) stackoverflow post where the author suggests to use `nginx-proxy`'s nginx instance, are there any repercussions to doing it like this?

**Comment by tkw1536** [comment]
You should probably use [Per Virtual-Host Configuration](https://github.com/nginx-proxy/nginx-proxy#per-virtual_host) config for this.
